File: app/src/main/java/ua/hospes/rtm/data/cars/CarsRepositoryImpl.java
package ua.hospes.rtm.data.cars;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.reactivex.Single;
import ua.hospes.rtm.data.cars.mapper.CarsMapper;
import ua.hospes.rtm.data.cars.storage.CarsDbStorage;
import ua.hospes.rtm.domain.cars.CarsRepository;
import ua.hospes.rtm.domain.cars.models.Car;
import ua.hospes.rtm.utils.RxUtils;

/**
 * @author Andrew Khloponin
 */
@Singleton
public class CarsRepositoryImpl implements CarsRepository {
    private final CarsDbStorage carsDbStorage;


    @Inject
    CarsRepositoryImpl(CarsDbStorage carsDbStorage) {
        this.carsDbStorage = carsDbStorage;
    }


    @Override
    public Observable<Car> get() {
        return carsDbStorage.get().map(CarsMapper::map);
    }

    @Override
    public Observable<Car> get(int... ids) {
        return carsDbStorage.get(ids).map(CarsMapper::map);
    }

    @Override
    public Observable<Car> getByNumbers(int... numbers) {
        return carsDbStorage.getByNumbers(numbers).map(CarsMapper::map);
    }

    @Override
    public Observable<Car> getNotInRace() {
        return carsDbStorage.getNotInRace().map(CarsMapper::map);
    }

    @Override
    public Observable<List<Car>> listen() {
        return carsDbStorage.listen().map(carDbs -> RxUtils.listMap(carDbs, CarsMapper::map));
    }

    @Override
    public Observable<Boolean> save(Car car) {
        if (car.getId() == -1) {
            return carsDbStorage.add(CarsMapper.map(car)).map(result -> result.getResult() != 0);
        } else {
            return carsDbStorage.update(CarsMapper.map(car)).map(result -> result.getResult() != 0);
        }
    }

    @Override
    public Single<Integer> remove(Car car) {
        return carsDbStorage.remove(CarsMapper.map(car));
    }

    @Override
    public Single<Integer> removeAll() {
        return carsDbStorage.removeAll();
    }
}

File: app/src/main/java/ua/hospes/rtm/data/cars/CarsRepositoryImpl.kt
package ua.hospes.rtm.data.cars

import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.functions.Function
import ua.hospes.rtm.domain.cars.Car
import ua.hospes.rtm.domain.cars.CarsRepository
import ua.hospes.rtm.utils.RxUtils

class CarsRepositoryImpl(private val carsDbStorage: CarsDbStorage) : CarsRepository {

    override fun get(): Observable<Car> = carsDbStorage.get().map { CarsMapper.map(it) }

    override fun get(vararg ids: Int): Observable<Car> = carsDbStorage.get(*ids).map { CarsMapper.map(it) }

    override fun getNotInRace(): Observable<Car> = carsDbStorage.getNotInRace().map { CarsMapper.map(it) }

    override fun listen(): Observable<List<Car>> = carsDbStorage.listen()
            .map { carDbs -> RxUtils.listMap<Car, CarDb>(carDbs, Function { CarsMapper.map(it) }) }

    override fun save(car: Car): Completable = when (car.id) {
        null -> carsDbStorage.add(CarsMapper.map(car)).map { result -> result.result != 0L }.toList().ignoreElement()
        else -> carsDbStorage.update(CarsMapper.map(car)).map { result -> result.result != 0L }.toList().ignoreElement()
    }

    override fun remove(id: Int): Completable = carsDbStorage.remove(id).ignoreElement()

    override fun removeAll(): Completable = carsDbStorage.removeAll().ignoreElement()
}